### PR TITLE
feat: implement forward token smearing

### PR DESCRIPTION
**Pull request by shehab-ashraf**
## Summary
Implements Prior Token Smearing to accelerate model convergence. This technique allows each token to "see" the immediately previous token by blending a small, gated amount of the prior embedding into the current one.
By connecting neighbors directly (token[t] += strength * token[t-1]), we prevent attention heads from wasting capacity on simple local lookups. This shortcut handles immediate context cheaply, allowing the expensive attention mechanism to focus its "brain power" on complex, long-range patterns.

### Smearing Direction: 
Previous / Prior ($t \leftarrow t-1$)
Current Token ($t$) blends with Prior Token ($t-1$).
First token ($t=0$) remains unchanged.

### Components:
smear_gate: 
Linear layer ($12 \to 1$). Uses only first 12 dims for efficiency.
smear_strength: 
Learnable scalar ($\lambda$). Initialized to 0.0 (starts as standard Transformer, learns to smear).

## Token Smearing

Let's say we have 3 tokens with embeddings of dimension 768 (but we'll show just 12 dims for clarity):

Before Smearing
```
Token 0: x0 = [1.2, 0.5, -0.3, 0.8, 1.1, -0.2, 0.4, 0.9, -0.5, 0.1, 0.7, -0.4, ...]  (768 dims total)
Token 1: x1 = [0.3, -0.2, 0.9, 1.4, -0.1, 0.6, -0.8, 0.2, 0.5, -0.3, 0.1, 0.8, ...]
Token 2: x2 = [-0.5, 0.8, 0.1, -0.3, 0.7, 1.2, 0.4, -0.6, 0.3, 0.9, -0.1, 0.2, ...]
```
### Step 1: Compute Gate Values

We use only the first 12 dimensions to decide how much to blend:
Why only 12 dims? Efficiency! We don't need all 768 dims to make a gating decision. 12 is enough to capture token characteristics while keeping computation cheap.

```
# For token 1
gate_input_1 = x1[0:12] = [0.3, -0.2, 0.9, 1.4, -0.1, 0.6, -0.8, 0.2, 0.5, -0.3, 0.1, 0.8]
# Small linear layer: 12 dims → 1 scalar (unnormalized)
gate_raw_1 = Linear(12, 1)(gate_input_1)  
# Let's say output = 0.3 (can be any value -∞ to +∞)
# Sigmoid: converts to 0..1 range
gate_1 = sigmoid(0.3) = 0.574   # now bounded between 0 and 1
# For token 2  
gate_input_2 = x2[0:12] = [-0.5, 0.8, 0.1, -0.3, 0.7, 1.2, 0.4, -0.6, 0.3, 0.9, -0.1, 0.2]
gate_2 = sigmoid(Linear(12, 1)(gate_input_2))
# Let's say = 0.421
```

Why sigmoid?  Linear layer can output any value (-∞ to +∞), but we want a gate in range 0..1:
0 = don't blend at all
1 = maximum blend
0.5 = medium blend

### Step 2: Apply Global Strength

```
smear_lambda -> LEARNABLE parameter (starts at 0)
strength_1 = smear_lambda * gate_1 = 0.07 * 0.574 = 0.040  # ~4% of previous token
strength_2 = smear_lambda * gate_2 = 0.07 * 0.421 = 0.029  # ~3% of previous token
```

Why multiply by smear_lambda? This scales down the gate values even more:
Gate gives 0..1 (per-token decision: "how much to blend?")
Lambda gives 0..0.07 (global limit: "don't blend too much!")

### Step 3: Blend with Previous Token

Now we add a weighted version of the previous token to current token
Token 1 gets blended with Token 0:
```
x1_new = x1 + strength_1 * x0
x1_new[0] = 0.3 + 0.040 * 1.2  = 0.3 + 0.048 = 0.348
x1_new[1] = -0.2 + 0.040 * 0.5 = -0.2 + 0.020 = -0.180
x1_new[2] = 0.9 + 0.040 * (-0.3) = 0.9 - 0.012 = 0.888
x1_new[3] = 1.4 + 0.040 * 0.8  = 1.4 + 0.032 = 1.432
... (all 768 dimensions)
x1_new = [0.348, -0.180, 0.888, 1.432, -0.056, 0.592, -0.784, 0.236, ...]
```

### After Smearing

```
Token 0: x0_new = [1.2, 0.5, -0.3, 0.8, 1.1, -0.2, 0.4, 0.9, -0.5, 0.1, 0.7, -0.4, ...]  (unchanged)
Token 1: x1_new = [0.348, -0.180, 0.888, 1.432, -0.056, 0.592, -0.784, 0.236, ...]  (has ~4% of x0)
Token 2: x2_new = [-0.491, 0.794, 0.126, -0.271, ...]  (has ~3% of x1)
```

## Implementation :
```
gate_input = x[:, 1:, :12].detach()
gates = sigmoid(Linear(gate_input))
strength = lambda * gates
x = torch.cat([
    x[:, :1],  
    x[:, 1:] + (strength * x[:, :-1].detach())
], dim=1)
```

### CODE REVIEW

**Comment by vukrosic** [comment]
@shehab-ashraf i see you are making changes, tell me when it's ready
